File: Lab_04/lang-tests/test-8/pasaje.c
#include <stdio.h>

// C por defecto tiene pasaje por valor, el pasaje por referencia
// se consigue pasando explicitamente las dirrecciones de memoria
// de dichos parametros.

// VAMOS A HACER UN PROGRAMA QUE DE OPCION DE ELEGIR EL PASAJE DE PARAMETRO A USAR


void interchange_reference(int *x1, int *y1) {
    int z1;

    z1 = *x1;
    *x1 = *y1;
    *y1 = z1;
}


void interchange_value(int x1, int y1) {
    int z1;

    z1 = x1;
    x1 = y1;
    y1 = z1;
}


int main() {
    int x = 50, y = 70;

    int change;

    printf("\nIngrese '0' si quiere Pasaje por Valor\nIngrese '1' si quiere Pasaje por Referencia\nIngrese la opcion: ");
    scanf("%d", &change);

    printf("\nANTES DE interchange: x = %d, y = %d\n", x, y);

    if (change == 0) {
        interchange_value(x, y);
    }
    else if (change == 1) {
        interchange_reference(&x, &y);
    }

    printf("\nDESPUES DE interchange: x = %d, y = %d\n\n", x, y);

    return 0;
}

File: Lab_04/lang-tests/test-8/pasaje.pl
# El pasaje de parametros de Perl es por REFERENCIA

$x = 50;
$y = 70;

print "\nANTES DE INTERCHANGE: x = $x, y = $y\n";

&interchange ($x, $y);

print "\nDESPUES DE INTERCHANGE: x = $x, y = $y\n\n";


sub interchange
{
    ($x1, $y1) = @_; # x1 e y1 son variables locales que toman x e y respectivamente

    @_[0] = $y1;

    @_[1] = $x1;

    #print "\nADENTRO DE INTERCHANGE (ARREGLO): x1 = @_[0], y1 = @_[1]\n"
    print "\nADENTRO DE INTERCHANGE (VALORES): x1 = $x1, y1 = $y1\n"

}